### ISSUE TITLE
Validate Foreign Keys across Data Packages

### ISSUE THREAD

**Issue by Stephen-Gates** [ODIQueensland/data-curator#277]
### Desired Behaviour (for feature requests only)

- implement pattern at https://frictionlessdata.io/specs/patterns/#table-schema:-foreign-keys-to-data-packages
- see discussion https://gitter.im/frictionlessdata/chat?at=59eaed08f7299e8f53142845 
- test data https://github.com/Stephen-Gates/data-package-examples/tree/master/donations
- note property name may be `package` rather than `datapackage` https://gitter.im/frictionlessdata/chat?at=59eeda7f210ac26920235bf0

**Comment by Stephen-Gates** [comment]
/remind me on april 12 to write the acceptance test for foreign keys

**Comment by Stephen-Gates** [comment]
[Acceptance test](https://app.cucumber.pro/projects/data-curator/documents/branch/develop/test/features/tools/validate-table.feature)

**Comment by Stephen-Gates** [comment]
Note: Some [debate](https://github.com/frictionlessdata/specs/issues/597#issuecomment-377739252) on json tag to describe the data package - `datapackage` or `package`?

**Comment by Stephen-Gates** [comment]
FYI https://gitter.im/frictionlessdata/chat?at=5ac6e1cd2dfcfffd2b2c2caf

**Comment by Stephen-Gates** [comment]
[Conversation continues](https://gitter.im/frictionlessdata/chat?at=5ad4bc1315c9b0311401783b) re: `datapackage` or `package`

**Comment by Stephen-Gates** [comment]
Going with `Package` see https://gitter.im/frictionlessdata/chat?at=5ad6b53701a2b40f382edca2

I’ll put in a pr for the pattern

**Comment by mattRedBox** [comment]
Hi @Stephen-Gates 

For foreign keys, here the 'package' is a URL.
For the validation to occur, am I correct in assuming that:
- need to check URL is valid
- need to check that URL exists (points to a 'live' resource)
Once has correct URL for 'package'
are the available resources reference table and reference column(s) to be downloaded and filled in the relevant places in Data-Curator?
or are these entered in free-text-fields for validation to check that these values are correct and do exist in 'package' URL?

**Comment by Stephen-Gates** [comment]
I didn’t envisage the table at the URL being downloaded for display in Data Curator.

**Comment by mattRedBox** [comment]
Thanks @Stephen-Gates 
Clears much of that up for me

**Comment by mattRedBox** [comment]
Implemented view for foreign keys from URL and alignment with local foreign keys.
Yet to add validation.

**Comment by Stephen-Gates** [comment]
@mattRedBox what do you think about adding the validation to v0.16.0 given the potential delay in contributing the code?

**Comment by mattRedBox** [comment]
There's no harm in trying :)
For the validation itself: I've just raised this issue to ensure I have the requirements covered:
https://github.com/frictionlessdata/datapackage-js/issues/103

**Comment by Stephen-Gates** [comment]
Note: PR to use `package` instead or `datapackage` has been [merged](https://github.com/frictionlessdata/specs/pull/611)

**Comment by mattRedBox** [comment]
Hi @Stephen-Gates 
In the example you've shown in [datapackage-js issue](https://github.com/frictionlessdata/datapackage-js/issues/103) the reason why you see that error is that the code is looking for a local table from which to create foreign key objects from Reference Table and Reference Column and cannot find any.
However going through this again, I'm struggling to understand where we implement code here. Atm when we validate, while we check for significant package properties, we use tableschema-js to read each table, including the table's foreign keys. As package properties doesn't have foreign keys, we don't use datapackage-js.

While atm the datapackage-js shows text input, I was thinking that with this piece of work we would actually load datapackage and from that, autofill with select dialogs (similar to what we have for local) to select only legitimate tables and headers. We would use the same code then to create the objects needed to send to the reading of foreign keys, which exists in tableschema-js. When we use tableschema-js read in our validation call, the relations objects are used to check that a table is referencing values that exist in the other table. I'm not sure how any of this changes with a datapackage URL added. If we're trying to validate the package URL, what exactly is it that we're validating? Is it that it is a legitimate datapackge? Or is there something more we're checking?

We already have code which loads a datapackage URL, which uses frictionless' datapackage.load, which can load a URL already, and use the existing frictionless call to check it is valid or has errors. We don't need to validate this again as once we've loaded the URL to fill tables and headers, we've already done this by initially loading the datapackage in a table's foreign keys (ie: if it's not legitimate we won't be able to pull down tables/headers to fill).

**Comment by Stephen-Gates** [comment]
Hi Matt, the screenshot there was for context only. Maybe we can tweak the error message if required after this change is implemented.

I agree fetching the resources and fields from the package url and using those values to seed to the dropdown boxes is a good idea.

I'm not sure I'm following everything you're saying. My thought was that after the user has entered:

- a url
- selected a resource 
- selected one or more fields in that resource
- pressed validate

then tableschema.js is called to validate the foreignKeys relationship only (given that we've already validated the url, package, and have valid resource and fields)

I thought the bulk of this change would be getting tableschema.js to validate the foreignKeys relationship.

Does that make sense???

**Comment by mattRedBox** [comment]
Hi @Stephen-Gates 
Yes, which is where I thought we would be putting a PR - I'm happy to go through and put in where I think the code update should be made and submit to as a PR in tableschema-js, but the link highlighted for where to start is in datapackage-js - which is where I was thrown, as I don't see how we would be using that codebase.

**Comment by louisjasek** [comment]
@Stephen-Gates @mattRedBox should I be testing this or are you still nutting it out? I had a test of the current functionality and had some trouble, so happy to leave you with it until it is resolved.

**Comment by Stephen-Gates** [comment]
@louisjasek the user interface is implemented but the backend isn’t. No need to test yet

**Comment by Stephen-Gates** [comment]
@mattRedBox test data with two tables https://github.com/frictionlessdata/example-data-packages/tree/master/countries-and-currencies

**Comment by mattRedBox** [comment]
OK
so atm can only fetch resource from another url
so will put in check to also guess relative path from resource and create url from this.

**Comment by mattRedBox** [comment]
Hi @Stephen-Gates 
I'm not quite sure why (and need to move on for now),
but when I download json as URL:
this package json works fine:
https://raw.githubusercontent.com/frictionlessdata/example-data-packages/master/geo-location-fk/datapackage.json

but this one fails when datapackage-js is trying to read the second resource.
https://raw.githubusercontent.com/frictionlessdata/example-data-packages/master/countries-and-currencies/datapackage.json

Just flagging for now - but might be something that we need to address later.

**Comment by Stephen-Gates** [comment]
> but this one fails when datapackage-js is trying to read the second resource.
https://raw.githubusercontent.com/frictionlessdata/example-data-packages/master/countries-and-currencies/datapackage.json

Not sure on that as the [data package is valid](https://datahub.io/tools/validate?q=https%3A%2F%2Fraw.githubusercontent.com%2Ffrictionlessdata%2Fexample-data-packages%2Fmaster%2Fcountries-and-currencies%2Fdatapackage.json)

**Comment by louisjasek** [comment]
This is working great for me. I have tried with a few different data packages and it works great. I also tried with multiple foreign keys and a combination of URL and local table. I note that data types have to match as well as the value (e.g. String of "1" would not match Integer of 1), which I think is good. For improvement, something to indicate there was an error loading the data package json (at the moment it just spins and then stops) - I'll add a comment to #651 - as Matt mentioned it may be possible to cover this in that issue.
